File: apps/mobile/shared/Sources/FlagshipCore/AdminRootRotation.swift
import Foundation
import CryptoKit

/// Slice D §5.1 — the admin master-root rotation proof. Swift mirror of
/// `AdminRootRotation` / `signAdminRootRotation` / `verifyAdminRootRotation`
/// in `packages/protocol/src/adminRootRotation.ts` (the spine's rotation
/// signer). The canonical bytes MUST match the TS engine byte-for-byte or
/// the box-side verify against its pinned `adminRootPub` fails.
///
/// The box must NOT trust `.com`'s word for a new admin authority root.
/// When credential recovery mints a fresh admin master root, the OLD admin
/// root signs an `AdminRootRotation{ old → new }`. The box verifies the
/// proof against its PINNED `adminRootPub` (the old root), then — and only
/// then — re-pins to the new root. `.com` relays the proof but can never
/// forge one (it lacks the old master root), which is exactly what lets a
/// box adopt a relayed new root.
///
/// Rotation EXCLUDES other admin devices: they still hold the OLD root, so
/// after the box re-pins to the new root their old-root-signed orders no
/// longer verify. That is the intended "revoke every other admin" semantic
/// of a compromise-recovery rotation (docs/device-admin-tier-spec.md §5).
///
/// Canonical bytes (field-guarded, `|`-separated), matching TS `hex()`
/// (lowercase) exactly:
///   flagship/admin-root-rotation/v1 | username | hex(oldAdminRootPub)
///     | hex(newAdminRootPub) | issuedAt
public struct AdminRootRotation: Equatable, Sendable {
    /// `flagship/admin-root-rotation/v1`, the same tag the spine uses.
    public static let canonicalTag = "flagship/admin-root-rotation/v1"

    public let username: String
    /// The box's currently-pinned admin root, lowercase hex (32 → 64 chars).
    public let oldAdminRootPubHex: String
    /// The freshly-minted admin master root the box re-pins to, lowercase hex.
    public let newAdminRootPubHex: String
    public let issuedAt: Int64

    public init(username: String, oldAdminRootPubHex: String, newAdminRootPubHex: String, issuedAt: Int64) {
        self.username = username
        self.oldAdminRootPubHex = oldAdminRootPubHex
        self.newAdminRootPubHex = newAdminRootPubHex
        self.issuedAt = issuedAt
    }

    /// `flagship/admin-root-rotation/v1|<username>|<oldPubHex>|<newPubHex>|<issuedAt>`.
    /// Mirrors the TS `legacyFieldGuard("username", …)`: the username may not
    /// carry the `|` separator (usernames never do — `^[a-z0-9][a-z0-9-]…$`).
    public func canonicalBytes() -> Data {
        Data(
            [
                Self.canonicalTag,
                username,
                oldAdminRootPubHex.lowercased(),
                newAdminRootPubHex.lowercased(),
                String(issuedAt),
            ].joined(separator: "|").utf8
        )
    }

    /// Sign with the OLD admin master root (the anchor the box already pins).
    public func sign(withOldAdminRoot key: Curve25519.Signing.PrivateKey) throws -> Data {
        try key.signature(for: canonicalBytes())
    }

    /// Verify the proof against the OLD admin master root pubkey (the pinned
    /// anchor). Returns false (never throws) on a malformed key/signature,
    /// mirroring the TS `verifyAdminRootRotation`.
    public func verify(signature: Data, oldAdminRootPub: Data) -> Bool {
        guard let pub = try? Curve25519.Signing.PublicKey(rawRepresentation: oldAdminRootPub) else {
            return false
        }
        return pub.isValidSignature(signature, for: canonicalBytes())
    }
}

File: packages/protocol/src/adminRootRotation.ts
/**
 * Slice D — admin master-root rotation proof (docs/device-admin-tier-spec.md §5).
 *
 * The box must NOT trust `.com`'s word for a new admin authority root. When
 * credential recovery mints a fresh admin master root, the OLD admin root signs
 * an `AdminRootRotation{ old → new }`. The box verifies the proof against its
 * PINNED `adminRootPub` (the old root), then — and only then — re-pins to the
 * new root. `.com` relays the proof but can never forge one (it lacks the old
 * master root), which is exactly what lets the box adopt a relayed new root.
 *
 * Canonical bytes (field-guarded, `|`-separated):
 *   flagship/admin-root-rotation/v1 | username | hex(oldAdminRootPub)
 *     | hex(newAdminRootPub) | issuedAt
 *
 * Signed by the OLD admin master root:
 *   signAdminRootRotation(r, oldAdminRoot)   → 64-byte Ed25519 sig
 *   verifyAdminRootRotation(r, sig, oldAdminRootPub)
 */
import { ed } from "./edSync.js";
import { hex, legacyFieldGuard } from "./canonicalBase.js";
import type { Bytes, Keypair } from "./types.js";

const TAG_ADMIN_ROOT_ROTATION = "flagship/admin-root-rotation/v1";

export interface AdminRootRotation {
  /** Account whose admin authority root is rotating. */
  username: string;
  /** MUST equal the box's currently-pinned adminRootPub (the anchor being replaced). */
  oldAdminRootPub: Bytes;
  /** The freshly-minted admin master root the box re-pins to on a valid proof. */
  newAdminRootPub: Bytes;
  /** ms since epoch. */
  issuedAt: number;
}

function canonicalAdminRootRotation(r: AdminRootRotation): Bytes {
  legacyFieldGuard("username", r.username);
  return new TextEncoder().encode(
    [
      TAG_ADMIN_ROOT_ROTATION,
      r.username,
      hex(r.oldAdminRootPub),
      hex(r.newAdminRootPub),
      r.issuedAt,
    ].join("|"),
  );
}

/** Sign with the OLD admin master root (the anchor the box already pins). */
export function signAdminRootRotation(r: AdminRootRotation, oldAdminRoot: Keypair): Bytes {
  return ed.sign(canonicalAdminRootRotation(r), oldAdminRoot.privateKey);
}

/** Verify the proof against the OLD admin master root pubkey (the pinned anchor). */
export function verifyAdminRootRotation(
  r: AdminRootRotation,
  sig: Bytes,
  oldAdminRootPub: Bytes,
): boolean {
  try {
    return ed.verify(sig, canonicalAdminRootRotation(r), oldAdminRootPub);
  } catch {
    return false;
  }
}
